Add response statistics (#6869)
triton-inference-server/server: docs/protocol/extension_statistics.md

  "inference_count" : $number,
  "execution_count" : $number,
  "inference_stats" : $inference_stats,

>   "response_stats" : { $string : $response_stat, ... },
>   "batch_stats" : [ $batch_stat, ... ],

Reviewer: ```suggestion
  "response_stats" : { $string : $response_stats, ... },
  "batch_stats" : [ $batch_stats, ... ],
```

  DOES NOT include cache hits.

- "inference_stats" : The aggregate statistics for the
  model. So, for example, "inference_stats":"success" indicates the number of
  successful inference requests for the model.

- "response_stats" : The aggregate decoupled response statistics for the

>   model. For example, { "key" : { "response_stats" : "success" } } indicates the

Reviewer 2: Can we do a real example here?  `key` is still confusing. Something like:
Assume there were `N` number of requests sent to server. An example of a possible response statistics is ...,
where <actual value> is a `key`, which helps to differentiate between decoupled responses generated by the model. It is advised ...

Author: Sure, added a brief example for key.

  DOES NOT include cache hits.

- "inference_stats" : The aggregate statistics for the
  model. So, for example, "inference_stats":"success" indicates the number of
  successful inference requests for the model.

> - "response_stats" : The aggregate decoupled response statistics for the

Reviewer: ```suggestion
- "response_stats" : The aggregate response statistics for the
```

Reviewer: would replace mentions of "decoupled response" to just "response"

Author: Sure, they are all updated on `docs/protocol/extension_statistics.md`.

  model. For example, { "key" : { "response_stats" : "success" } } indicates the
  aggregate statistics of successful decoupled responses at "key" for the model,
  where "key" identifies each decoupled response generated by the model across

>   different requests. For example, given a model that generates three responses,

Reviewer: I think it makes sense to call out the example model "decoupled' in this case since it is described to have multiple responses
```suggestion
  different requests. For example, given a decoupled model that generates three responses,
```

  the key could be "0", "1" and "2" identifying the three responses in order.

>   Check with the model backend on the keys it can return and their meanings.

Reviewer: The "keys" aren't backend specific are they? Isn't it a fixed set defined in core/proto?

Author: They are not backend specific, at this point, unless we allow customizable keys for the backend in the future. I only left the example behind, I think that should be enough.

  model/version. So, for example, "inference_stats":"success"
  indicates the number of successful inference requests for the model.

- "response_stats" : The aggregate decoupled response statistics for the
  model/version. For example, { "key" : { "response_stats" : "success" } }
  indicates the aggregate statistics of successful decoupled response at "key"
  for the model, where "key" identifies between different decoupled responses

>   generated by the model.

Reviewer 2: > The aggregate decoupled response statistics for the model/version.

I think we need to clarify `model/version`. Is it the specific version of a specific model?

>  indicates the aggregate statistics of successful decoupled response at "key" for the model

I am a little bit confused what `at "key" for the model` should mean.

>   where "key" identifies between different decoupled responses generated by the model.

` identifies between different decoupled responses` <- not sure what you mean by this.

Author: > I think we need to clarify model/version. Is it the specific version of a specific model?

It is a specific version of a specific model, because the backend views different versions of a model different models. Let's use "model" for it, because except one or two other places, the rest of the docs refers to "model".

> I am a little bit confused what at "key" for the model should mean.

It identifies between different decoupled responses generated by the model. For example, a model generates 3 responses, the key can be "0", "1" and "2". 

> identifies between different decoupled responses <- not sure what you mean by this.

Basically a response, even an empty one, should be mapped to a uniquely identifiable value -> "key". It is ultimately the backend responsibility for supplying valid keys for grouping response statistics. i.e. send the response statistics at the right order to ensure it is mapped to the correct index ("0", "1", "2", ...).

I added a comment below to ask the user to check with the backend on the set of keys and what each means.

  // point, the GPU memory usage for models in ONNX Runtime backend and TensorRT
  // backend is usually aligned.
  repeated MemoryUsage memory_usage = 8;

>   // The key and value pairs for all decoupled responses statistics. The key is

Reviewer: ```suggestion
  // The key and value pairs for all response statistics. The key is
```

  StatisticDuration cache_miss = 8;
}

> // Statistics per decoupled response.

Reviewer: ```suggestion
// Statistics per response.
```
